### PR TITLE
`AttrDict` raises `AttributeError` on missing attributes, not `KeyError`.

### PR DESCRIPTION
**Pull request by craigcitro**
Why? Consistency.

As noted in [the Python docs](https://docs.python.org/3/reference/datamodel.html#object.__getattr__), a custom `__getattr__` method should raise `AttributeError` when an attribute isn't present. However, the existing `scipy.io.idl.AttrDict` class sets `__getattr__ = __getitem__`, leading to a `KeyError` in this case.

The result is that these objects behave in unexpected ways; a concrete example where this came up is https://github.com/googlecolab/colabtools/issues/2472. This commit switches to instead raising an `AttributeError` in this situation.

* Before: [`KeyError`](https://user-images.githubusercontent.com/468559/144984611-a5b38e3e-fd25-4ae1-a29e-cb28808dca26.png)
* After: [`AttributeError`](https://user-images.githubusercontent.com/468559/144984626-95f7cc14-7b35-4afc-b7d6-0024125df013.png)

### CODE REVIEW

**Comment by WarrenWeckesser** [comment]
Agreed, attempting to read an attribute that doesn't exist should raise an `AttributeError`.

This might be considered a backwards-incompatible change, but I view it as a bug fix.

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
Thanks @craigcitro,  the addition of `__getattr__` looks good.  My suggested changes are only in the docstring and the tests.

It would be good to get the opinion of another dev on the change of the error type.  As I said earlier, this looks like a bug fix to me, so  I think the change is fine. (A short release note about the change could be added if the PR is merged.)

File: scipy/io/_idl.py
Comment (WarrenWeckesser) [reply scipy/io/_idl.py:652]: On second thought, this is fine.  When I wrote the above comment, I was thinking this was in the `readsav` docstring.  As part of the `AttrDict` docstring, I guess it is OK.

Comment (craigcitro) [reply scipy/io/_idl.py:652]: FWIW I also went back and forth on adding these to the docstring -- they're probably overkill, but there's no other documentation for `AttrDict`, so I figured it didn't hurt.

I'll leave them in for now, but happy to drop them if it feels too verbose.

File: scipy/io/tests/test_idl.py
Comment (craigcitro) [reply scipy/io/tests/test_idl.py:448]: Done.

**Comment by craigcitro** [comment]
Thanks @WarrenWeckesser!

Re: "get the opinion of another dev" -- is there somewhere I should ask for someone else to chime in?

**Comment by rgommers** [comment]
> It would be good to get the opinion of another dev on the change of the error type.

I agree with this change.

**Review by WarrenWeckesser — APPROVED** [review]
Looks good, thanks @craigcitro.

**Comment by craigcitro** [comment]
Awesome, thanks @WarrenWeckesser and @rgommers! Anything else I need to do or just wait for someone with commit rights to merge?

**Comment by WarrenWeckesser** [comment]
Merged.  Thanks @craigcitro.

**Pull request merged**
